### PR TITLE
Refactor Python binding

### LINKED ISSUE #785: `flexflow_python` built in `python` instead of the build directory
@eddy16112 Why is this currently the behavior, and can it be easily fixed? It can be rather annoying when two build directories with different settings are used, and each has to be rebuilt every time because they overwrite each others' `flexflow_python`

### PR DESCRIPTION
**Pull request by eddy16112**
**Description of changes:**

This is a big refactor of python binding. We do not build our own `flexflow_python`, instead, we rely on the `legion_python` provided by Legion. There are two major benefits:
1. We can seamlessly switch between legion python and native python interpreter. If we run 
```
legion_python flexflow_script.py
```
FlexFlow will detect that legion python is used, and use legion python, if we run
```
python flexflow_script.py
```
FlexFlow will detect that legion python is not used, so fall back to use native python. 
 
2. FlexFlow now will work with other Legion python libraries such as Legate. 

TODOs:

- [x] fix makefile
- [x] factor the cpp code such that we generate a single libflexflow.so instead of a `libflexflow.so` and `libflexflow_c.so`
- [x] fix the initialization of FlexFlow::FFConfig. Currently we get machine configuration such as number of gpus from argv. Since we are now register tasks and mapper after runtime is started, we are not able to get machine configurations from argv because once Legion consumes `-ll:gpu`, it will be removed from argv, and FlexFlow can not get the value from the argv. 

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #785 

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

**Comment by eddy16112** [comment]
Both CMake and Makefile work now, we just need to fix the `make install` for CMake, feel free to take a look at it.

File: python/main.cc [hunk 1]
```diff
@@ -1,106 +0,0 @@
...
-      setenv("PYTHONPATH", buffer, true /*overwrite*/);
-    } else {
-      setenv("PYTHONPATH", bin_dir, true /*overwrite*/);
-    }
-
-    free(bin_path);
-  }
-#endif
-
-#ifdef FF_USE_NCCL
```
Comment (gabrieleoliaro) [inline python/main.cc:80]: @eddy16112 @jiazhihao do you think the lack of this block of code in the new interface might be related to the nccl-related segfault we're seeing?

Comment (gabrieleoliaro) [reply python/main.cc:80]: https://flexflowgroup.slack.com/archives/C046R49TYQK/p1677192959436439

Comment (eddy16112) [reply python/main.cc:80]: We do have the nccl code here https://github.com/flexflow/FlexFlow/blob/python_refactor/src/c/flexflow_c.cc#L1920, not sure if it is too late. One possible solution to try it to move the nccl code before https://gitlab.com/StanfordLegion/legion/-/blob/master/bindings/python/main.cc#L137 
or put `env["NCCL_LAUNCH_MODE"] = "PARALLEL"` before https://github.com/flexflow/FlexFlow/blob/python_refactor/python/flexflow/core/__init__.py#L25

**Comment by lockshaw** [comment]
@eddy16112 @gabrieleoliaro This closes #71, correct? If so, can we add it under the closed issues in the PR description?

**Comment by lockshaw** [comment]
@eddy16112 What's the status of this PR? Is it blocked on a different issue/PR?

**Review by goliaro — APPROVED** [review]

**Review by goliaro — CHANGES_REQUESTED** [review]
Right now, everything seems to work if we apply the following changes to Legion: [see here](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838/diffs). To get all tests to work in this PR, we will have to either [merge PR 838](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838) in Legion, or change our cmake files and tests to always install FlexFlow, instead of building it and running in place. This is because `bindings/python/CMakeLists.txt` does not currently have code to build the Legion Python bindings without installing them

**Comment by eddy16112** [comment]
> Right now, everything seems to work if we apply the following changes to Legion: [see here](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838/diffs). To get all tests to work in this PR, we will have to either [merge PR 838](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838) in Legion, or change our cmake files and tests to always install FlexFlow, instead of building it and running in place. This is because `bindings/python/CMakeLists.txt` does not currently have code to build the Legion Python bindings without installing them

Can we do make install? Legion's cmake will install the legion files into the conda env.

**Comment by goliaro** [comment]
> > Right now, everything seems to work if we apply the following changes to Legion: [see here](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838/diffs). To get all tests to work in this PR, we will have to either [merge PR 838](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838) in Legion, or change our cmake files and tests to always install FlexFlow, instead of building it and running in place. This is because `bindings/python/CMakeLists.txt` does not currently have code to build the Legion Python bindings without installing them
> 
> Can we do make install? Legion's cmake will install the legion files into the conda env.

I think so! But why should we change FlexFlow so that its python bindings are only built if we install it somewhere, and remove the option to run FlexFlow without installing? With the changes in the `bindings/python/CMakeLists.txt` we can always run `make install` and install everything into the conda env as you said, but we can also build in place and run FlexFlow, with or without conda. We can also build multiple versions of FlexFlow on the same machine (e.g. when working on several projects on different branches) without having to create and activate a separate conda environment for each of them

**Comment by eddy16112** [comment]
> > > Right now, everything seems to work if we apply the following changes to Legion: [see here](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838/diffs). To get all tests to work in this PR, we will have to either [merge PR 838](https://gitlab.com/StanfordLegion/legion/-/merge_requests/838) in Legion, or change our cmake files and tests to always install FlexFlow, instead of building it and running in place. This is because `bindings/python/CMakeLists.txt` does not currently have code to build the Legion Python bindings without installing them
> > 
> > 
> > Can we do make install? Legion's cmake will install the legion files into the conda env.
> 
> I think so! But why should we change FlexFlow so that its python bindings are only built if we install it somewhere, and remove the option to run FlexFlow without installing? 

I think your PR just runs setup.py as the post make operation, which is the same as make install. You are just make the `make` as `make install`, there is no difference.

**Comment by jiazhihao** [comment]
Let's discuss this PR during the FF meeting today.

**Review by goliaro — APPROVED** [review]

**Pull request merged**